File: android/src/main/java/com/margelo/nitro/pdfeditor/HybridPdfDocument.kt
package com.margelo.nitro.pdfeditor

import com.facebook.proguard.annotations.DoNotStrip
import com.margelo.nitro.core.ArrayBuffer
import com.margelo.nitro.core.Promise
import com.podofo.android.PdfDocument as PodofoDocument
import com.podofo.android.PdfPermission as PodofoPermission

private fun PdfPermissions?.toPodofoBitmask(): Int {
  if (this == null) return PodofoPermission.DEFAULT
  var result = 0
  if (print != false) result = result or PodofoPermission.PRINT
  if (edit != false) result = result or PodofoPermission.EDIT
  if (copy != false) result = result or PodofoPermission.COPY
  if (editNotes != false) result = result or PodofoPermission.EDIT_NOTES
  if (fillAndSign != false) result = result or PodofoPermission.FILL_AND_SIGN
  if (accessible != false) result = result or PodofoPermission.ACCESSIBLE
  if (docAssembly != false) result = result or PodofoPermission.DOC_ASSEMBLY
  if (highPrint != false) result = result or PodofoPermission.HIGH_PRINT
  return result
}

/**
 * Runs [block] holding both [lockA] and [lockB], always acquired in the same global order (by
 * identity hash) regardless of which document called which
 * - avoids deadlock if two documents are merged into each other concurrently from different
 *   threads. Mirrors the iOS C++ side's `std::scoped_lock` in `HybridPdfDocument::appendPagesFrom`
 *   and friends.
 */
private fun <T> synchronizedBoth(lockA: Any, lockB: Any, block: () -> T): T {
  val (first, second) =
    if (System.identityHashCode(lockA) <= System.identityHashCode(lockB)) {
      lockA to lockB
    } else {
      lockB to lockA
    }
  return synchronized(first) { synchronized(second) { block() } }
}

private fun Standard14FontName.toPodofoName(): String =
  when (this) {
    Standard14FontName.TIMESROMAN -> "TimesRoman"
    Standard14FontName.TIMESITALIC -> "TimesItalic"
    Standard14FontName.TIMESBOLD -> "TimesBold"
    Standard14FontName.TIMESBOLDITALIC -> "TimesBoldItalic"
    Standard14FontName.HELVETICA -> "Helvetica"
    Standard14FontName.HELVETICAOBLIQUE -> "HelveticaOblique"
    Standard14FontName.HELVETICABOLD -> "HelveticaBold"
    Standard14FontName.HELVETICABOLDOBLIQUE -> "HelveticaBoldOblique"
    Standard14FontName.COURIER -> "Courier"
    Standard14FontName.COURIEROBLIQUE -> "CourierOblique"
    Standard14FontName.COURIERBOLD -> "CourierBold"
    Standard14FontName.COURIERBOLDOBLIQUE -> "CourierBoldOblique"
    Standard14FontName.SYMBOL -> "Symbol"
    Standard14FontName.ZAPFDINGBATS -> "ZapfDingbats"
  }

/**
 * Wraps the podofo-android JNI wrapper's com.podofo.android.PdfDocument. See PLAN.md §2: unlike iOS
 * (which binds Nitro's C++ layer directly to PoDoFo's core), the published podofo-android AAR only
 * exposes this compiled Java wrapper — no headers/static libs for direct linkage.
 *
 * [lock] guards every native call into this document's underlying object graph — the podofo-android
 * wrapper classes (PdfDocument/PdfPage/PdfPainter/PdfField/PdfAnnotation) have no internal
 * synchronization of their own (confirmed by reading their source: no `synchronized` anywhere), and
 * `save()` runs on a background thread via `Promise.parallel` while the document remains fully
 * callable from JS. Shared (not copied) with every HybridPdfPage/HybridPdfPainter/
 * HybridPdfField/HybridPdfAnnotation obtained from this document — mirrors the iOS C++ side's
 * `std::mutex` fix in HybridPdfDocument.hpp/cpp.
 */
@DoNotStrip
class HybridPdfDocument(private val native: PodofoDocument) : HybridPdfDocumentSpec() {
  private val lock = Any()

  override val pageCount: Double
    get() = synchronized(lock) { native.pageCount.toDouble() }

  override fun getPage(index: Double): HybridPdfPageSpec {
    return synchronized(lock) { HybridPdfPage(native.getPage(index.toInt()), lock) }
  }

  override fun createPage(width: Double, height: Double): HybridPdfPageSpec {
    return synchronized(lock) { HybridPdfPage(native.createPage(width, height), lock) }
  }

  override fun createPageAt(index: Double, width: Double, height: Double): HybridPdfPageSpec {
    return synchronized(lock) {
      HybridPdfPage(native.createPageAt(index.toInt(), width, height), lock)
    }
  }

  override fun removePageAt(index: Double) {
    synchronized(lock) { native.removePageAt(index.toInt()) }
  }

  override fun appendPagesFrom(source: HybridPdfDocumentSpec) {
    val src = source as HybridPdfDocument
    require(src !== this) { "Cannot merge a document into itself" }
    synchronizedBoth(lock, src.lock) { native.appendPagesFrom(src.native) }
  }

  override fun appendPageRangeFrom(
    source: HybridPdfDocumentSpec,
    pageIndex: Double,
    pageCount: Double,
  ) {
    val src = source as HybridPdfDocument
    require(src !== this) { "Cannot merge a document into itself" }
    synchronizedBoth(lock, src.lock) {
      native.appendPagesFrom(src.native, pageIndex.toInt(), pageCount.toInt())
    }
  }

  override fun insertPageFrom(atIndex: Double, source: HybridPdfDocumentSpec, pageIndex: Double) {
    val src = source as HybridPdfDocument
    require(src !== this) { "Cannot merge a document into itself" }
    synchronizedBoth(lock, src.lock) {
      native.insertPageFrom(atIndex.toInt(), src.native, pageIndex.toInt())
    }
  }

  override fun getStandard14Font(name: Standard14FontName): HybridPdfFontSpec {
    return synchronized(lock) { HybridPdfFont(native.getStandard14Font(name.toPodofoName())) }
  }

  override fun loadFont(path: String): HybridPdfFontSpec {
    return synchronized(lock) { HybridPdfFont(native.getOrCreateFont(path)) }
  }

  override fun loadFontFromBuffer(data: ArrayBuffer): HybridPdfFontSpec {
    return synchronized(lock) {
      HybridPdfFont(native.getOrCreateFontFromBuffer(data.toByteArray()))
    }
  }

  override fun createImageFromBuffer(data: ArrayBuffer): HybridPdfImageSpec {
    return synchronized(lock) { HybridPdfImage(native.createImageFromBuffer(data.toByteArray())) }
  }

  override fun save(path: String): Promise<Unit> {
    return Promise.parallel<Unit> { synchronized(lock) { native.save(path) } }
  }

  override fun getTitle(): String? = synchronized(lock) { native.title }

  override fun setTitle(title: String?) {
    synchronized(lock) { native.title = title }
  }

  override fun getAuthor(): String? = synchronized(lock) { native.author }

  override fun setAuthor(author: String?) {
    synchronized(lock) { native.author = author }
  }

  override fun getSubject(): String? = synchronized(lock) { native.subject }

  override fun setSubject(subject: String?) {
    synchronized(lock) { native.subject = subject }
  }

  override fun getCreator(): String? = synchronized(lock) { native.creator }

  override fun setCreator(creator: String?) {
    synchronized(lock) { native.creator = creator }
  }

  override val fieldCount: Double
    get() = synchronized(lock) { native.fieldCount.toDouble() }

  override fun getFieldAt(index: Double): HybridPdfFieldSpec {
    return synchronized(lock) { HybridPdfField(native.getFieldAt(index.toInt()), lock) }
  }

  override fun createTextBox(name: String): HybridPdfFieldSpec {
    return synchronized(lock) { HybridPdfField(native.createTextBox(name), lock) }
  }

  override fun createCheckBox(name: String): HybridPdfFieldSpec {
    return synchronized(lock) { HybridPdfField(native.createCheckBox(name), lock) }
  }

  override fun setEncrypted(
    userPassword: String,
    ownerPassword: String,
    permissions: PdfPermissions?,
  ) {
    synchronized(lock) {
      native.setEncrypted(userPassword, ownerPassword, permissions.toPodofoBitmask())
    }
  }

  override fun isEncrypted(): Boolean = synchronized(lock) { native.isEncrypted }

  override fun dispose() {
    synchronized(lock) {
      native.close()
    }
    super.dispose()
  }
}

File: cpp/HybridPdfDocument.cpp
#include "HybridPdfDocument.hpp"
#include "HybridPdfField.hpp"
#include "HybridPdfFont.hpp"
#include "HybridPdfImage.hpp"
#include "HybridPdfPage.hpp"

namespace margelo::nitro::pdfeditor {

using namespace PoDoFo;

static PdfStandard14FontType toPodofoStandard14(Standard14FontName name) {
  switch (name) {
    case Standard14FontName::TIMESROMAN:
      return PdfStandard14FontType::TimesRoman;
    case Standard14FontName::TIMESITALIC:
      return PdfStandard14FontType::TimesItalic;
    case Standard14FontName::TIMESBOLD:
      return PdfStandard14FontType::TimesBold;
    case Standard14FontName::TIMESBOLDITALIC:
      return PdfStandard14FontType::TimesBoldItalic;
    case Standard14FontName::HELVETICA:
      return PdfStandard14FontType::Helvetica;
    case Standard14FontName::HELVETICAOBLIQUE:
      return PdfStandard14FontType::HelveticaOblique;
    case Standard14FontName::HELVETICABOLD:
      return PdfStandard14FontType::HelveticaBold;
    case Standard14FontName::HELVETICABOLDOBLIQUE:
      return PdfStandard14FontType::HelveticaBoldOblique;
    case Standard14FontName::COURIER:
      return PdfStandard14FontType::Courier;
    case Standard14FontName::COURIEROBLIQUE:
      return PdfStandard14FontType::CourierOblique;
    case Standard14FontName::COURIERBOLD:
      return PdfStandard14FontType::CourierBold;
    case Standard14FontName::COURIERBOLDOBLIQUE:
      return PdfStandard14FontType::CourierBoldOblique;
    case Standard14FontName::SYMBOL:
      return PdfStandard14FontType::Symbol;
    case Standard14FontName::ZAPFDINGBATS:
      return PdfStandard14FontType::ZapfDingbats;
  }
  throw std::invalid_argument("Unknown Standard14FontName");
}

static std::optional<std::string> toOptionalString(
    const nullable<const PdfString&>& value) {
  if (!value.has_value()) {
    return std::nullopt;
  }
  return std::string(value->GetString());
}

static PoDoFo::PdfPermissions toPodofoPermissions(
    const std::optional<PdfPermissions>& permissions) {
  if (!permissions.has_value()) {
    return PoDoFo::PdfPermissions::Default;
  }
  PoDoFo::PdfPermissions result = PoDoFo::PdfPermissions::None;
  const auto& p = *permissions;
  if (p.print.value_or(true))
    result |= PoDoFo::PdfPermissions::Print;
  if (p.edit.value_or(true))
    result |= PoDoFo::PdfPermissions::Edit;
  if (p.copy.value_or(true))
    result |= PoDoFo::PdfPermissions::Copy;
  if (p.editNotes.value_or(true))
    result |= PoDoFo::PdfPermissions::EditNotes;
  if (p.fillAndSign.value_or(true))
    result |= PoDoFo::PdfPermissions::FillAndSign;
  if (p.accessible.value_or(true))
    result |= PoDoFo::PdfPermissions::Accessible;
  if (p.docAssembly.value_or(true))
    result |= PoDoFo::PdfPermissions::DocAssembly;
  if (p.highPrint.value_or(true))
    result |= PoDoFo::PdfPermissions::HighPrint;
  return result;
}

double HybridPdfDocument::getPageCount() {
  std::lock_guard<std::mutex> lock(*_mutex);
  return static_cast<double>(_doc->GetPages().GetCount());
}

std::shared_ptr<HybridPdfPageSpec> HybridPdfDocument::getPage(double index) {
  std::lock_guard<std::mutex> lock(*_mutex);
  auto& page = _doc->GetPages().GetPageAt(static_cast<unsigned>(index));
  return std::make_shared<HybridPdfPage>(_doc, _mutex, &page);
}

std::shared_ptr<HybridPdfPageSpec> HybridPdfDocument::createPage(
    double width, double height) {
  std::lock_guard<std::mutex> lock(*_mutex);
  auto& page = _doc->GetPages().CreatePage(Rect(0, 0, width, height));
  return std::make_shared<HybridPdfPage>(_doc, _mutex, &page);
}

std::shared_ptr<HybridPdfPageSpec> HybridPdfDocument::createPageAt(
    double index, double width, double height) {
  std::lock_guard<std::mutex> lock(*_mutex);
  auto& page = _doc->GetPages().CreatePageAt(static_cast<unsigned>(index),
                                             Rect(0, 0, width, height));
  return std::make_shared<HybridPdfPage>(_doc, _mutex, &page);
}

void HybridPdfDocument::removePageAt(double index) {
  std::lock_guard<std::mutex> lock(*_mutex);
  _doc->GetPages().RemovePageAt(static_cast<unsigned>(index));
}

void HybridPdfDocument::appendPagesFrom(
    const std::shared_ptr<HybridPdfDocumentSpec>& source) {
  auto& src = static_cast<HybridPdfDocument&>(*source);
  if (&src == this) {
    throw std::invalid_argument("Cannot merge a document into itself");
  }
  // Locks both documents (ordered internally by std::scoped_lock to avoid
  // deadlock if two documents are merged into each other concurrently from
  // different threads) since this reads the source's page tree while
  // writing to this document's.
  std::scoped_lock lock(*_mutex, *src._mutex);
  _doc->GetPages().AppendDocumentPages(*src._doc);
}

void HybridPdfDocument::appendPageRangeFrom(
    const std::shared_ptr<HybridPdfDocumentSpec>& source, double pageIndex,
    double pageCount) {
  auto& src = static_cast<HybridPdfDocument&>(*source);
  if (&src == this) {
    throw std::invalid_argument("Cannot merge a document into itself");
  }
  std::scoped_lock lock(*_mutex, *src._mutex);
  _doc->GetPages().AppendDocumentPages(*src._doc,
                                       static_cast<unsigned>(pageIndex),
                                       static_cast<unsigned>(pageCount));
}

void HybridPdfDocument::insertPageFrom(
    double atIndex, const std::shared_ptr<HybridPdfDocumentSpec>& source,
    double pageIndex) {
  auto& src = static_cast<HybridPdfDocument&>(*source);
  if (&src == this) {
    throw std::invalid_argument("Cannot merge a document into itself");
  }
  std::scoped_lock lock(*_mutex, *src._mutex);
  _doc->GetPages().InsertDocumentPageAt(static_cast<unsigned>(atIndex),
                                        *src._doc,
                                        static_cast<unsigned>(pageIndex));
}

std::shared_ptr<HybridPdfFontSpec> HybridPdfDocument::getStandard14Font(
    Standard14FontName name) {
  std::lock_guard<std::mutex> lock(*_mutex);
  // PdfFontManager (PdfDocument::GetFonts()) owns and caches standard-14
  // fonts itself — no need to keep our own reference alive (confirmed
  // against the podofo repo's own working JNI wrapper implementation,
  // src/wrapper/podofo_jni.cpp, which uses this same call).
  auto& font = _doc->GetFonts().GetStandard14Font(toPodofoStandard14(name));
  return std::make_shared<HybridPdfFont>(&font);
}

std::shared_ptr<HybridPdfFontSpec> HybridPdfDocument::loadFont(
    const std::string& path) {
  std::lock_guard<std::mutex> lock(*_mutex);
  // PdfFontManager caches by path and owns the result itself (same as
  // GetStandard14Font above) — repeated calls with the same path are cheap
  // and return the same underlying font.
  auto& font = _doc->GetFonts().GetOrCreateFont(path);
  return std::make_shared<HybridPdfFont>(&font);
}

std::shared_ptr<HybridPdfFontSpec> HybridPdfDocument::loadFontFromBuffer(
    const std::shared_ptr<ArrayBuffer>& data) {
  std::lock_guard<std::mutex> lock(*_mutex);
  auto& font = _doc->GetFonts().GetOrCreateFontFromBuffer(
      bufferview(reinterpret_cast<const char*>(data->data()), data->size()));
  return std::make_shared<HybridPdfFont>(&font);
}

std::shared_ptr<HybridPdfImageSpec> HybridPdfDocument::createImageFromBuffer(
    const std::shared_ptr<ArrayBuffer>& data) {
  std::lock_guard<std::mutex> lock(*_mutex);
  auto image = _doc->CreateImage();
  image->LoadFromBuffer(
      bufferview(reinterpret_cast<const char*>(data->data()), data->size()));
  PdfImage* rawImage = image.get();
  _images.push_back(std::move(image));
  return std::make_shared<HybridPdfImage>(rawImage);
}

std::shared_ptr<Promise<void>> HybridPdfDocument::save(
    const std::string& path) {
  auto doc = _doc;
  auto mutex = _mutex;
  return Promise<void>::async([doc, mutex, path]() {
    std::lock_guard<std::mutex> lock(*mutex);
    doc->Save(path);
  });
}

std::optional<std::string> HybridPdfDocument::getTitle() {
  std::lock_guard<std::mutex> lock(*_mutex);
  return toOptionalString(_doc->GetMetadata().GetTitle());
}

void HybridPdfDocument::setTitle(const std::optional<std::string>& title) {
  std::lock_guard<std::mutex> lock(*_mutex);
  _doc->GetMetadata().SetTitle(
      title.has_value() ? nullable<const PdfString&>(PdfString(*title))
                        : nullptr);
}

std::optional<std::string> HybridPdfDocument::getAuthor() {
  std::lock_guard<std::mutex> lock(*_mutex);
  return toOptionalString(_doc->GetMetadata().GetAuthor());
}

void HybridPdfDocument::setAuthor(const std::optional<std::string>& author) {
  std::lock_guard<std::mutex> lock(*_mutex);
  _doc->GetMetadata().SetAuthor(
      author.has_value() ? nullable<const PdfString&>(PdfString(*author))
                         : nullptr);
}

std::optional<std::string> HybridPdfDocument::getSubject() {
  std::lock_guard<std::mutex> lock(*_mutex);
  return toOptionalString(_doc->GetMetadata().GetSubject());
}

void HybridPdfDocument::setSubject(const std::optional<std::string>& subject) {
  std::lock_guard<std::mutex> lock(*_mutex);
  _doc->GetMetadata().SetSubject(
      subject.has_value() ? nullable<const PdfString&>(PdfString(*subject))
                          : nullptr);
}

std::optional<std::string> HybridPdfDocument::getCreator() {
  std::lock_guard<std::mutex> lock(*_mutex);
  return toOptionalString(_doc->GetMetadata().GetCreator());
}

void HybridPdfDocument::setCreator(const std::optional<std::string>& creator) {
  std::lock_guard<std::mutex> lock(*_mutex);
  _doc->GetMetadata().SetCreator(
      creator.has_value() ? nullable<const PdfString&>(PdfString(*creator))
                          : nullptr);
}

double HybridPdfDocument::getFieldCount() {
  std::lock_guard<std::mutex> lock(*_mutex);
  auto* form = _doc->GetAcroForm();
  return form == nullptr ? 0 : static_cast<double>(form->GetFieldCount());
}

std::shared_ptr<HybridPdfFieldSpec> HybridPdfDocument::getFieldAt(
    double index) {
  std::lock_guard<std::mutex> lock(*_mutex);
  auto& form = _doc->GetOrCreateAcroForm();
  auto& field = form.GetFieldAt(static_cast<unsigned>(index));
  return std::make_shared<HybridPdfField>(_doc, _mutex, &field);
}

std::shared_ptr<HybridPdfFieldSpec> HybridPdfDocument::createTextBox(
    const std::string& name) {
  std::lock_guard<std::mutex> lock(*_mutex);
  auto& form = _doc->GetOrCreateAcroForm();
  auto& field = form.CreateField(name, PoDoFo::PdfFieldType::TextBox);
  return std::make_shared<HybridPdfField>(_doc, _mutex, &field);
}

std::shared_ptr<HybridPdfFieldSpec> HybridPdfDocument::createCheckBox(
    const std::string& name) {
  std::lock_guard<std::mutex> lock(*_mutex);
  auto& form = _doc->GetOrCreateAcroForm();
  auto& field = form.CreateField(name, PoDoFo::PdfFieldType::CheckBox);
  return std::make_shared<HybridPdfField>(_doc, _mutex, &field);
}

void HybridPdfDocument::setEncrypted(
    const std::string& userPassword, const std::string& ownerPassword,
    const std::optional<PdfPermissions>& permissions) {
  std::lock_guard<std::mutex> lock(*_mutex);
  _doc->SetEncrypted(userPassword, ownerPassword,
                     toPodofoPermissions(permissions));
}

bool HybridPdfDocument::isEncrypted() {
  std::lock_guard<std::mutex> lock(*_mutex);
  return _doc->IsEncrypted();
}

}  // namespace margelo::nitro::pdfeditor
